### DISCUSSION TITLE
Making prediction of masked sequence

### DISCUSSION THREAD

**Discussion by frisbyts** [facebookresearch/esm#85] [Q&A] [answered]
Hello,

I would like to use the model to make predictions on masked sequences.  I've found that if I pass the model a sequence with `<mask>` in place of an amino acid, the model mistakes each individual character as an unknown rather than interpreting it as a masked position.  What is the correct way to indicate that an input sequence is masked?  Thanks.

**Comment by joshim5** [comment, marked as answer]
Hi @frisbyts, can you elaborate on what you mean by "mistakes each individual character as an unknown?" How do you how the model is interpreting this?

If you give us more information about what you have tried, then we can try to help.

**Reply by frisbyts** [reply]
Thank you for your quick reply!  I have a sequence of length 56.  Suppose I want to mask the second position.  I then have the following data:

`data = [("seq1","M<mask>YKLILNGKTLKGETTTEAVDAATAEKVFKQYANDNGVDGEWTYDDATKTFTVTE")]`

I want to provide this sequence to the model, so I call `batch_converter`:

`batch_labels, batch_strs, batch_tokens = batch_converter(data)`

If I inspect the batch tokens, I notice that the sequence has increased in length by 6 ignoring the start/end tokens (the length of the string `<mask>`), where each "masked" token has been assigned 3, which corresponds to `<unk>` in 'alphabet.all_toks':

`[ 0, 20,  3,  3,  3,  3,  3,  3, 19, 15,  4, 12,  4, 17,  6, 15, 11,  4,
   15,  6,  9, 11, 11, 11,  9,  5,  7, 13,  5,  5, 11,  5,  9, 15,  7, 18,
   15, 16, 19,  5, 17, 13, 17,  6,  7, 13,  6,  9, 22, 11, 19, 13, 13,  5,
    11, 15, 11, 18, 11,  7, 11,  9,  2]`

I understand that the model has not been invoked at this point, but I am clearly providing it with a sequence that does not represent what I want it to.  Am I mistaken in how I have masked the input sequence?

**Reply by tomsercu** [reply]
Hi, and sorry for the delay - the problem is that in BatchConverter, [iteration over the string](https://github.com/facebookresearch/esm/blob/master/esm/data.py#L201-L203) causes each character to be treated separately, ie "<", "m", etc, each of those are out of vocab.
Two easy solutions:
* provide the sequence as list `["M", "<mask>", "Y", "K" etc]`
* Mask after batch conversion, ie input as `"M-YK"` then for any position `pos_ix` in sequence `seq_ix` you set `batch_tokens[seq_ix, pos_ix] = alphabet.mask_idx`.

**Reply by tomsercu** [reply]
FYI We added an easier way to tokenize masked sequences, commit https://github.com/facebookresearch/esm/commit/3bb552af and demonstrated in https://github.com/facebookresearch/esm#quick-start-

**Comment by xf-omics** [comment]
I am interested in this question as well. I think you can input your sequence as 
"M-YKLILNGKTLKGETTTEAVDAATAEKVFKQYANDNGVDGEWTYDDATKTFTVTE".

From your tutorial, I see we can extract embedded representations from different layers. Can we also get the output (which amino acid is predicted to be the most possible one for each position)? Thanks.

**Reply by tomsercu** [reply]
See above, the `-` may be suboptimal since out of vocab, better to explicitly set the position to `alphabet.mask_idx` how the model was trained.

The logits are in `output["logits"]`, if you want probabilities of tokens you can do `probs = logits.softmax(-1)`.
